### ISSUE TITLE
Handle Racy Git Problem Cleanly

### ISSUE THREAD

**Issue by abraithwaite** [github/git-lfs#196]
This is less of an issue with `git-lfs` and more of an issue with git but users should still know about this.  It's something I learned while working on git-fat and it's very tricky to figure out.

[Git has a racy problem](https://github.com/git/git/blob/master/Documentation/technical/racy-git.txt).  It's manifested particularly well in filters.  [read-cache.c](https://github.com/git/git/blob/master/read-cache.c) has the code relating to this issue.

I won't go into all the detail that the excellent documentation does, but here's a tldr anyways:
git uses os.stat to determine whether or not a file needs to be updated in the index or the working tree.  If the timestamp from os.stat is the same for the working copy and the index, then it [compares both files](https://github.com/git/git/blob/master/read-cache.c#L157).  The problem you may now see is that the working copy and the index are the same and therefore won't do a smudge on the git-lfs file.

I tested it on git-lfs and saw the same behavior I saw in git-fat which I just described (asdf.zip is an empty file which matches the gitattributes, which is why cat shows nothing): https://gist.github.com/abraithwaite/6ed377b56b9fc23c2e13

The way we handled this in git-fat was providing a [`git fat checkout`](https://github.com/cyaninc/git-fat/blob/develop/git_fat/git_fat.py#L819-L846) command which manually went through and updated all the orphans which were found, if any.  Of course, the race condition still exists, so we remove the file if we find an orphan in the index and the working copy.

You can then run this command manually or as a `post-checkout` hook.

Please let me know if you have any questions or you need me to explain it better.  :racehorse:

**Comment by tclem** [comment]
Very interesting. Thanks for writing this up @abraithwaite.

**Comment by mhagger** [comment]
@abraithwaite: Thanks for the heads-up about this problem. I'm struggling to understand whether this is a Git bug, or whether there is something intrinsic to the way that filters work that makes your `git fat checkout` command necessary.

Part of my difficulty understanding your report is that your Gist example has left out a few steps. Was `asdf.zip` already added at the beginning of the transcript? What is the difference between branch `qwer` and branch `master`? It would help if you presented the *whole* test transcript, starting with `git init`.

**Comment by abraithwaite** [comment]
I'll give a more concrete example this after I get off work.  The problem is somewhat abated by the fact that you return non-zero from the filter if you can't download the file, but it leaves the repository in an unusable state.

e.g. Try interrupting a download during a clone and inspect the state of the repository (if you're feeling ambitious, otherwise wait and I'll include it in my example).

**Comment by technoweenie** [comment]
FWIW there's some more recent discussion at #1195. Also just merged the ability to stop returning non zero exit codes on download errors in #1213. 

TLDR, I think there are some changes that Git could potentially make. Until then, you might be better off using LFS's skip-smudge feature (`git lfs install --skip-smudge`), which requires you to run `git lfs pull` manually after every checkout. I'd love to use `post-checkout`, but it's not called in some very important cases (`git reset`, `git pull`, `git merge`, `git rebase`).

Closing this since it's ancient.

**Comment by thcipriani** [comment]
Here's a simple reproduction case.

1. Add `git-lfs` your path
2. Run this:
```bash
#!/usr/bin/env bash
sleep_to_next_second(){
    sleep 0.$(printf '%04d' $((10000 - 10#$(date +%4N))))
}

REPO=racy-git

# It's important that this all run within a single second
sleep_to_next_second
git clone https://github.com/thcipriani/racy-git.git "$REPO"
git lfs install
git -C "$REPO" checkout -f HEAD
```

When you check the `large-binary-file.bin` there:
```
root@a92f706a9e11:~/racy-git# cat large-binary-file.bin 
version https://git-lfs.github.com/spec/v1
oid sha256:98ea6e4f216f2fb4b69fff9b3a44842c38686ca685f3f55dc48c5d3fb1107be4
size 3
```

And it's only fixable either via modifying the mtime on the file, or running `git lfs pull`

```
root@9b4af497b45c:~/racy-git# cat large-binary-file.bin 
version https://git-lfs.github.com/spec/v1
oid sha256:98ea6e4f216f2fb4b69fff9b3a44842c38686ca685f3f55dc48c5d3fb1107be4
size 3
root@9b4af497b45c:~/racy-git# git checkout-index --index --force large-binary-file.bin 
root@9b4af497b45c:~/racy-git# cat large-binary-file.bin 
version https://git-lfs.github.com/spec/v1
oid sha256:98ea6e4f216f2fb4b69fff9b3a44842c38686ca685f3f55dc48c5d3fb1107be4
size 3
root@9b4af497b45c:~/racy-git# touch large-binary-file.bin 
root@9b4af497b45c:~/racy-git# git checkout-index --index --force large-binary-file.bin 
Downloading large-binary-file.bin (3 B)
root@9b4af497b45c:~/racy-git# cat large-binary-file.bin 
hi
```

It's definitely a git filter problem.

**Comment by bk2204** [comment]
I don't see this behaviour on my system, which is using Git 2.42 on Ubuntu 23.10.  What operating system and Git version are you using?

**Comment by thcipriani** [comment]
To be clear: this is not a problem I expect people to run into.

There are many ways to work around this problem as a user. And git-lfs is excellent at handling edge-cases like this (except this weird one I found).

But this is a clearer demonstration of the problem (I believe) the original poster was alluding to in this task.

The problem is the smudge filter. It's not being triggered by `git checkout -f HEAD` because the mtime and size of the files in the worktree match what's in the index.

This could be fixed, but it'd be tricky. And maybe the fix is even in git itself.

Details to recreate below

* * *

> I don't see this behaviour on my system, which is using Git 2.42 on Ubuntu 23.10. What operating system and Git version are you using?

- git version 2.25.1
- Ubuntu 20.04

Here's a Dockerfile that can recreate the problem:
```
FROM ubuntu:20.04

RUN apt-get update && apt-get install -y \
    git \
    git-lfs \
    && rm -rf /var/lib/apt/lists/*

ENV GIT_CONFIG_NOSYSTEM=1
COPY racy-git.sh /racy-git.sh
ENTRYPOINT ["/racy-git.sh"]
```

`GIT_CONFIG_NOSYSTEM=1` is here to mimic the case where `git-lfs` has not been installed yet.

The `racy-git.sh` script is:

```
#!/usr/bin/env bash
REPO=racy-git
git clone https://github.com/thcipriani/racy-git.git "$REPO"
git lfs install
git -C "$REPO" checkout -f HEAD
```

This should show you the problem:
```
docker build -t racy-git -f Dockerfile .
docker run --rm -it --entrypoint /bin/bash racy-git
bash /racy-git.sh
```

**Comment by abraithwaite** [comment]
The fix that needs to happen in git itself is a patch which allows the user to filter out fetching objects over a certain size, if it's not in the current working tree.  Similar to a shallow clone but for objects themselves instead of the commit/tree.

It would also be possible to resolve + download those large binaries at runtime when running a checkout, similar to the smudge filters work but happening before the checkout is resolved, rather than after.

This patch would completely obsolete git-lfs, git-fat and the handful of other competing largefile managers.  I don't have faith that github would pay to have someone do this, because the implications of such a change are probably nuanced and many (which is why I never attempted it).  Even if they are in the best position to get it merged, it's misaligned with their business incentives and probably would not be easy.

**Comment by bk2204** [comment]
I think what you're proposing is that `git checkout -f` should always check out all files in the working tree and always invoke smudge and clean filters.  If so, that's not what it's intended or documented to do, and it's not a bug in either Git or Git LFS that it has that behaviour.

**Comment by thcipriani** [comment]
> I think what you're proposing is that `git checkout -f` should always check out all files in the working tree and always invoke smudge and clean filters. If so, that's not what it's intended or documented to do, and it's not a bug in either Git or Git LFS that it has that behaviour.

That is not what I'm proposing—I'm not proposing anything.

I only came to document behavior that surprised me as a user:
- I added a filter driver
- ran a command which triggers filters
- the filter didn't run

My intuition was: adding a new `filter.<driver>` to the config should  trigger the filter for paths which specify that filter driver in a repo's attributes.

But this could be: not a bug/PBKAC.

I realize why it happens given the way git decides whether to run filters based on its index.

"Fixing" this may be a lot of work (modifying the index to track a portion of gitconfig and attributes...maybe(?)) and I'm clueless to the knock-on effects of such a change—it may result in much more unintuitive behavior/terrible performance.

Happy to leave it to you all to triage this :)

**Comment by abraithwaite** [comment]
> This exists and is called partial clone. It's not a perfect solution, either. It's also not on topic for this issue, which is discussing a totally different problem.

Thanks, I didn't know about partial clones!  🎉 

As the OP, I didn't think it was off topic, but apologies anyway.  Hopefully someone else with a similar problem might discover partial clones from this thread and find them a useful tool to address their problem in the future. :-)
